x-pack/plugins/encrypted_saved_objects/server/create_model_version.ts

```diff
@@ -0,0 +1,150 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { mergeTransformFunctions } from '@kbn/core-saved-objects-base-server-internal';
+import type {
+  SavedObjectModelUnsafeTransformFn,
+  SavedObjectsModelChange,
+  SavedObjectsModelUnsafeTransformChange,
+  SavedObjectsModelVersion,
+} from '@kbn/core-saved-objects-server';
+
+import { EncryptionError } from './crypto';
+import type { EncryptedSavedObjectsService, EncryptedSavedObjectTypeRegistration } from './crypto';
+import { mapAttributes } from './saved_objects/map_attributes';
+
+export interface CreateEsoModelVersionFnOpts {
+  modelVersion: SavedObjectsModelVersion;
+  shouldTransformIfDecryptionFails?: boolean;
+  inputType?: EncryptedSavedObjectTypeRegistration;
+  outputType?: EncryptedSavedObjectTypeRegistration;
+}
+
+// This function is designed to wrap a Model Version implementation of an Encrypted Saved Object (a Saved Object
+// who's type is registered with the Encrypted Saved Object Plugin). The purpose of this wrapper is to ensure that
+// version changes to the ESO what would require re-encryption (e.g.changes to encrypted fields or fields excluded
+// from AAD) are performed correctly. Prior to Model Versions, the CreateEncryptedSavedObjectsMigrationFn handled
+// wrapping migration functions for the same purpose.
+//
+// For Model Versions, 'unsafe_transform' changes are leveraged to implement any changes that require re-encryption.
+// This funtion returns a Model Version where the 'unsafe_transform' changes are merged into a single transform where
+// the document being transformed is first decrypted via the inputType EncryptedSavedObjectTypeRegistration (or by
+// default, the EncryptedSavedObjectTypeRegistration registered with the ESO Plugin), then transformed based on the
+// 'unsafe_transform' changes defined in the the input Model Versions, and finally encrypred via the outputType
+// EncryptedSavedObjectTypeRegistration (or by default, the EncryptedSavedObjectTypeRegistration registered with the
+// ESO Plugin). The implementation for this can be found in getCreateEsoModelVersion below.
+export type CreateEsoModelVersionFn = (
+  opts: CreateEsoModelVersionFnOpts
+) => SavedObjectsModelVersion;
+
+export const getCreateEsoModelVersion =
+  (
+    encryptedSavedObjectsService: Readonly<EncryptedSavedObjectsService>,
+    instantiateServiceWithLegacyType: (
+      typeRegistration: EncryptedSavedObjectTypeRegistration
+    ) => EncryptedSavedObjectsService
+  ): CreateEsoModelVersionFn =>
+  ({ modelVersion, shouldTransformIfDecryptionFails, inputType, outputType }) => {
+    // If there are no unsafe changes, then there is no reason to create an Encrypted Saved Objects Model Version
+    // Throw an error to notify the developer
+    const incomingUnsafeChanges = modelVersion.changes.filter(
+      (change) => change.type === 'unsafe_transform'
```

Comment:
So, at this point, I wonder why we're only supporting `unsafe_transform` type of changes.

Given you're importing from protected Core code anyway (`mergeTransformFunctions`), why just not import `buildModelVersionTransformFn` instead (from the same file), that would allow to simply and automatically support all other (and future) type of MV change types?

https://github.com/elastic/kibana/blob/52eb488692ca74cdc388188100540cc39e5d887e/packages/core/saved-objects/core-saved-objects-base-server-internal/src/model_version/build_transform_fn.ts#L18-L37

The ESO wrapping logic would be the same as what's done in this PR, no?

Unless I'm missing something? WDYT?

Reply:
> I was under the impression that we could only support unsafe_transform due to some raw document access limitation

Both `mergeTransformFunctions` and `buildModelVersionTransformFn` returns the same type (`SavedObjectModelTransformationFn`), so I think we should be fine?

Reply:
I just pushed a change to support all transform types.

Reply:
Ack, will TAL today